R/crunch-data-frame.R

```diff
@@ -0,0 +1,291 @@
+                             categorical.mode = "factor",
+                             include.hidden = FALSE) {
+    ## S3 constructor method for CrunchDataFrame. terms.formula doesn't seem
+    ## to like S4 subclass of environment
+    stopifnot(is.dataset(dataset))
+    
+    if (include.hidden) {
+        var_names <- aliases(allVariables(dataset))
+    } else {
+        var_names <- aliases(variables(dataset))
+    }
+    
+    out <- new.env()
+    attr(out, "crunchDataset") <- dataset
+    attr(out, "col_names") <- var_names
+    attr(out, "mode") <- categorical.mode
+    
+    with(out, {
+        ## Note the difference from as.environment: wrapped in as.vector
+        for (.a in var_names) {
+            eval(substitute(delayedAssign(v, {
+                getVarFromServer(col_name = v, crdf = ds, mode = mode)
+            }),
+            list(v=.a, ds=out, mode = attr(out, "mode"))))
+        }
+    })
+    
+    # set the order of the dataset based on row.order, if no row.order is 
+    # given return all rows in the dataset in the order they appear
+    attr(out, "order") <- row.order
+    
+    # TODO: add "data.frame" here too
+    class(out) <- "CrunchDataFrame"
+    return(out)
+}
+
+setOldClass("CrunchDataFrame")
+
+#' @export
+#' @rdname CrunchDataFrame
+dim.CrunchDataFrame <- function (x) {
+    if (is.null(attr(x, "order"))) {
+        # if there is no ordering, the num of rows is the same as dataset
+        n_rows <- nrow(attr(x, "crunchDataset"))
+    } else {
+        # if there is an ordering, the num of rows is the length of the ordering
+        n_rows <- length(attr(x, "order"))
+    }
+    return(c(n_rows, length(ls(x))))
+}
+
+#' @export
+#' @rdname CrunchDataFrame
+names.CrunchDataFrame <- function (x) attr(x, "col_names")
+
+
+
+#' @export
+#' @rdname CrunchDataFrame
+`[.CrunchDataFrame` <- function (x, i, j, drop = TRUE) {
+    j <- grabColNames(x, j)
+    row_inds <- grabRowInd(x, i)
+    
+    # grab the columns
+    out <- lapply(j, function(col) {
+        col_data <- x[[col]]
+        if (is.data.frame(col_data)) {
+            return(col_data[row_inds,])
+        } else {
+            return(col_data[row_inds])
+        }
+    })
+    
+    # set names, if j is a character use that, if not grab names from the crdf
+    if (is.character(j)) {
+        names(out) <- j
+    } else {
+        names(out) <- names(x)[j]
+    }
+    
+    if (length(out) == 1 & drop) {
+        # single column, so return vector
+        out <- out[[1]]
+    } else {
+        out <- structure(out, class="data.frame", row.names=seq_along(row_inds))
+    }
+    
+    return(out)
+}
+
+#' @export
+#' @rdname CrunchDataFrame
+`[<-.CrunchDataFrame` <- function (x, i, j, value) {
+    j <- grabColNames(x, j)
+    row_inds <- grabRowInd(x, i)
+    
+    # TODO: check value length
+    
+    # split value string by the length of row indicators
+    values <- split(value, ceiling(seq_along(value)/length(row_inds)))
```

Comment:
I'm not following this. Can you explain?

Reply:
That splits the values into a list that corresponds to each column. So it takes the values and splits them up into chunks that are the length of the number of rows that were specified. This is to match this pattern:

```
foo <- data.frame(foo=c(1, 2, 3, 4), bar=c(5, 6, 7, 8), baz = c(1, 2, 3, 4), qux = c(5, 6, 7, 8))
foo[c(1, 2), c(3, 4)] <- c(10, 20, 30, 40)
```

I should add in additional checks that ensure that there are exactly the right number of values  (currently that will break...)

Reply:
That's a really unnatural insertion. 

Can you please update the comment, perhaps using this example? "value string" isn't accurate, apparently, and "row indicators" is unclear. In reality, you're "splitting by row" because the base behavior assumes that you're inserting a matrix, which is column major.